pro_tes/middleware/middleware_pipeline.py

```diff
@@ -0,0 +1,127 @@
+"""Set up the middleware pipeline."""
+import importlib
+import logging
+
+from flask import current_app
+
+logger = logging.getLogger(__name__)
+
+
+class MiddlewarePipeline:
+    """Middleware pipeline."""
+
+    def __init__(self, middlewares=None):
+        """Construct object instance."""
+        self.middlewares = middlewares or []
+        logger.info(f"middleware: {self.middlewares}")
+
+    def add_middleware(self, middleware):
+        """Add a middleware to the pipeline."""
+        self.middlewares.append(middleware)
+
+    def process_request(self, request, *args, **kwargs):
+        """Process the incoming request.
+
+        Args:
+            request: Incoming request object.
+            *args: Additional positional arguments.
+            ***kwargs: Additional keyword arguments.
+
+        Returns:
+            The modified request object.
+        """
+        for main_middleware, fallback_middlewares in self.middlewares:
```

Comment:
I wouldn't even bother with distinguishing "main" from "fallback" middlwares. Just process the entire list (if it's a list) and stop when the first one succeeds. If it's just a string, then try that. If none works, raise an error.